QuestionAnswering/QA_ICL/README.md

```diff
@@ -1,15 +1,213 @@
+The Question-Answering System for The World Avatar involves retrieving data from RDF graphs and other data sources such as HTTP endpoints. To do so, input questions need to be converted into SPARQL queries and HTTP requests, whose execution would yield the desired data. The conversion of natural language queries to data requests is facilitated by in-context learning (ICL), which entails engineering a text prompt for Large Language Models (LLMs) to automatically perform the transformation. The prompt may include context information such as parsing examples and the structure of target predictions.
+
+Key Features
+
+- Utilises in-context learning for semantic parsing
+- Supports multiple data sources including RDF graphs and HTTP endpoints
+- Employs LLMs for natural language understanding
+- Provides a user-friendly frontend interface
+- Includes utilities for data preparation
+
+
+## Architecture
+
+1. Frontend: **Next.js app** serves as the web interface.
+2. Backend: 
+    - **FastAPI app**: handles requests from Next.js app with the help of other services.
+    - **Redis**: helps with
+      - search, including vector similarity and text-based search for retrieving LLM inference contexts and entity linking;
+      - expirable storage of QA requests and responses.
+    - **Triton server**: serves Sentence-BERT model for text embedding generation via gRPC.
+3. External services:
+   - **Blazegraph servers**: store and expose RDF data via SPARQL.
+   - **OpenAI API server**: provides chat completions API.
+
+```mermaid
+graph TD
+    subgraph Frontend
+    next[<b>Next.js app</b><br/>Web interface]
+    end
+
+    subgraph Backend
+    fastapi[<b>FastAPI app</b>]
+    redis[(<b>Redis server</b><br/>Search &<br/>expirable storage)]
+    triton[<b>Triton server</b><br/>Text embedding<br/>generation]
+    end
+
+    bg[(<b>Blazegraph servers</b><br/>RDF data storage)]
 
-## Architecture overview
+    openai[<b>OpenAI API server</b><br/>Chat completions]
+
+    next -->|HTTP| fastapi
+    fastapi -->|RESP| redis
+    fastapi -->|gRPC| triton
+    fastapi -->|SPARQL| bg
+    fastapi -->|HTTP| openai
+
+    note(Arrow direction and label denotes dependency and the protocol of communication)
+    style note stroke-dasharray: 5 5
+```
 
-![](uml/system_architecture.png)
 
 ## Project Structure
 
-- [`data_generation`](./data_generation): encompases Python scripts to prepare data for entity linking and in-context learning.
-- [`backend`](./backend): Marie's backend
-- [`frontend`](./frontend): Next app for Marie frontend and a minimal mock backend server
```

Comment:
I think it is a step backwards to remove these hyperlinks to the data generation, backend, and frontend readmes. These are _very_ convenient for clicking through the documentation! Also, as it stands, these readmes are now quite repetitive/redundant/duplicating content. Is there a plan to resolve this?

Reply:
I have added the hyperlinks to the new project structure section.

Regarding the setup steps, the plan for writing the README files is as follows. The idea is to describe the native installation (for development) and Docker setup (for deployment) only once for each component. The only exception is the root README, which gives a rundown of everything needed to deploy Marie with NGINX reverse proxy. It is intended that this README can be read independently from other sets of installation instructions.
- data_generation/README.md: 
  - Python env setup
- backend/
  - fastapi_app/README.md: 
    - Native setup
    - Docker setup
  - triton_inference_server/README.md: 
    - Docker setup
  - README.md:
    - Docker setup, referencing the instructions in the fastapi_app and triton_inference_server subdirectories
- frontend/
  - next_app_marie:
    - Native setup
    - Docker setup
  - mock_backend:
    - Native setup
- README.md (root): a single set of deployment instructions using Docker, together with NGINX configuration.

Is there any other areas of repetition/redundance/duplication in the README files that you would like to be addressed?

Reply:
I've addressed the following: 
1. Adding a "How to Use This Documentation" section to guide users on effective navigation.
1. Adding a clarifying statement about the independence of the main README in the Production deployment guide section.
2. Adding a descriptions  in the Project Structure section and links to component-specific READMEs.